beancount.el

```diff
@@ -570,30 +578,40 @@ With an argument move to the previous non cleared transaction."
                     (complete-with-action action candidates string pred))))
             (list (match-beginning 1) (match-end 1) completion-table))))))))
 
-(defun beancount-collect (regexp n)
-  "Return an unique list of REGEXP group N in the current buffer."
+(defun beancount-collect-pos-alist (regexp n)
+  "Return a list of conses mapping matches of REGEXP group N in the
+current buffer to a position of the match beginning."
   (let ((pos (point)))
     (save-excursion
       (save-match-data
-        (let ((hash (make-hash-table :test 'equal)))
+        (let (result)
           (goto-char (point-min))
           (while (re-search-forward regexp nil t)
             ;; Ignore matches around `pos' (the point position when
             ;; entering this funcyion) since that's presumably what
             ;; we're currently trying to complete.
             (unless (<= (match-beginning 0) pos (match-end 0))
-              (puthash (match-string-no-properties n) nil hash)))
-          (hash-table-keys hash))))))
+              (push (cons (match-string-no-properties n) (match-beginning 0))
+                    result)))
+          (nreverse result))))))
```

Comment:
The reason why `beancount-collect` used an hash table is that the amount of postings in a ledger can be huge, thus accumulating all the account names ever occurring in the ledger easily build ups to a very very long list. Creating a huge list for then immediately dropping most of it in a de-duplication step didn't seem very efficient way, especially because this function is called for completion, thus latency matter. Have you thought about the performance implications?

Reply:
Yes, and this works for completion as it only needs unique postings. But looking for references means that we have to go through all of the account postings.

So we have to cache unique posting names separately for completion purposes, and build a full list of references separately.

Reply:
I understand why the change is made, no problem with that. I'm just worried that the change may make completion slower for very large ledgers. However, I don't have a very good grasp for elisp performance, thus this may be premature optimization.

One possible solution to keep the optimization and avoid code duplication, would be to make `beancount-collect` get a reference to a function to call for each match found and pass in different functions for the two use cases.

Reply:
@dnicolodi I've just checked a relatively big file (25k lines). On my 2-3 years old notebook no noticeable there was no noticeable slowdown for definition search and completion.

This is becase the list is only built once and then a cached deduplicated list is used. It's the reference search with =M-?= that results in 0.5-1 sec slowdown. 

Let me know if you think speeding things up is still worth the trouble.